packages/rest/src/router/routing-table.ts

```diff
@@ -328,20 +437,40 @@ export class ControllerRoute extends BaseRoute {
       args,
     );
   }
-
-  private async _createControllerInstance(
-    requestContext: Context,
-  ): Promise<ControllerInstance> {
-    const valueOrPromise = instantiateClass(
-      this._controllerCtor,
-      requestContext,
-    );
-    return (await Promise.resolve(valueOrPromise)) as ControllerInstance;
-  }
 }
 
 function describeOperationParameters(opSpec: OperationObject) {
   return ((opSpec.parameters as ParameterObject[]) || [])
     .map(p => (p && p.name) || '')
     .join(', ');
 }
+
+/**
+ * Create a controller factory function
+ * @param source The source can be one of the following:
+ * - A binding key
+ * - A controller class
+ * - A controller instance
+ */
+export function createControllerFactory<T>(
+  source: ControllerClass<T> | string | T,
+): ControllerFactory<T> {
+  if (typeof source === 'string') {
+    return ctx => ctx.get<T>(source);
+  } else if (typeof source === 'function') {
```

Comment:
These function is doing sort of two things: provide a string-based and class-based factory. I'd like it more to extract these two factories into standalone functions and require users to pick the right factory function when creating a controller route (see my previous comment about a required `controllerFactory` arg), instead of letting the framework to create the factory function automagically. When I write "users", it does not necessary mean framework users building applications. I can imagine the high-level application API providing the factory function for callers, but keeping our internal design cleaner by making the factory function required.

I would be ok with leaving this out of the scope of this pull request, if it was not affecting the public API in a backwards-incompatible way. If backwards compatibility is not a concern, then I am ok with landing the currently proposed API.

Thoughts?

Reply:
I split it into three specific functions